rids/rules.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2022 Jigsaw Operations LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Generalizes the various formats of IOCs into Rule and RuleSet.
+
+Rule is a value type that represents evaluation on a single rule.  RuleSet
+instances represent a set of independent and overlapping Rule instances.
+
+RuleSet should typically be constructed from one of the flavors of parsers
+in `rids.iocs` or merged from combining a pair of RuleSet instances.  Its
+
+Evaluation of the RuleSet is done with the ProcessHandshake(...) and
+ProcessEndpoint(...) methods which take an observation and return a list
+of zero or more Event instances with their matching rules in the RuleSet.
+"""
+
+
+import collections
+from dataclasses import dataclass
+
+from rids import observations
+
+
+class RuleSet:
+    __slots__ = ('ip_address_rules', 'allowed_sni_port') 
+
+    def __init__(self):
+      self.ip_address_rules = collections.defaultdict(list)
+      # TODO this allow-list could be expressed in terms of rules from a file
+      # similar to the indicators of compromise, but explicitly coded for now.
+      self.allowed_sni_port = set([
```

Comment:
Create some `allow_sni_port()` so the class is reusable.
Put the list in some factory function you use for RIDS to create the suspicious tls object.
That way this rule set can be used in other contexts and you can write tests that are independent of the list we actually use.

Reply:
Since the PR was already around 700 lines I had thought to leave it for after the initial review.  The TODO here was about the kind of abstraction you suggest, and related to your suggestion elsewhere, I've made it so you can specify this list in the config.  Let me know what you think.